File: src/routes/repositories.js
const express = require('express');

const {
  createOrUpdateRepository,
  listRepositories,
  getRepositoryById,
  validateRepository,
  deleteRepository,
} = require('../services/repositories');

const router = express.Router();

router.get('/repositories', (req, res) => {
  res.json({
    repositories: listRepositories(),
  });
});

router.get('/repositories/:id', (req, res) => {
  const repository = getRepositoryById(req.params.id);

  if (!repository) {
    return res.status(404).json({
      error: 'Repository not found',
    });
  }

  return res.json(repository);
});

router.post('/repositories', (req, res) => {
  const {
    fullName,
    provider,
    localPath,
    defaultBranch,
    pipelineFile,
    active,
  } = req.body;

  const result = createOrUpdateRepository({
    fullName,
    provider,
    localPath,
    defaultBranch,
    pipelineFile,
    active,
  });

  if (!result.ok) {
    return res.status(400).json({
      error: result.reason,
    });
  }

  return res.status(201).json({
    message: 'Repository saved',
    repository: result.repository,
  });
});

router.post('/repositories/:id/validate', (req, res) => {
  const result = validateRepository(req.params.id);

  if (!result.ok) {
    const statusCode = result.reason === 'Repository not found' ? 404 : 400;

    return res.status(statusCode).json({
      error: result.reason,
    });
  }

  return res.json({
    message: 'Repository configuration is valid',
    repository: result.repository,
    pipelinePath: result.pipelinePath,
    pipeline: result.pipeline,
  });
});

router.delete('/repositories/:id', (req, res) => {
  const result = deleteRepository(req.params.id);

  if (!result.ok) {
    return res.status(404).json({
      error: result.reason,
    });
  }

  return res.json({
    message: 'Repository deleted',
    repository: result.repository,
  });
});

module.exports = router;

File: app/routers/repositories.py
from __future__ import annotations

from fastapi import APIRouter, HTTPException

from ..schemas import RepositoryCreate
from ..services.repositories import (
    create_or_update_repository,
    delete_repository,
    get_repository_by_id,
    list_repositories,
    validate_repository,
)

router = APIRouter(prefix="/repositories", tags=["repositories"])


@router.get("")
def repositories_list() -> dict:
    return {"repositories": list_repositories()}


@router.get("/{repo_id}")
def repositories_get(repo_id: str):
    repository = get_repository_by_id(repo_id)
    if not repository:
        raise HTTPException(status_code=404, detail="Repository not found")
    return repository


@router.post("", status_code=201)
def repositories_create(body: RepositoryCreate) -> dict:
    result = create_or_update_repository(**body.model_dump())
    if not result["ok"]:
        raise HTTPException(status_code=400, detail=result["reason"])
    return {"message": "Repository saved", "repository": result["repository"]}


@router.post("/{repo_id}/validate")
def repositories_validate(repo_id: str) -> dict:
    result = validate_repository(repo_id)
    if not result["ok"]:
        raise HTTPException(
            status_code=404 if result["reason"] == "Repository not found" else 400,
            detail=result["reason"],
        )
    return {
        "message": "Repository configuration is valid",
        "repository": result["repository"],
        "pipelinePath": result["pipelinePath"],
        "pipeline": result["pipeline"],
    }


@router.delete("/{repo_id}")
def repositories_delete(repo_id: str) -> dict:
    result = delete_repository(repo_id)
    if not result["ok"]:
        raise HTTPException(status_code=404, detail=result["reason"])
    return {"message": "Repository deleted", "repository": result["repository"]}
